src/agents/user_agent.py

```diff
@@ -1,12 +1,126 @@
-import spade
+import json
+from typing import List
 from spade.agent import Agent
-from spade.behaviour import CyclicBehaviour, OneShotBehaviour
+from spade.behaviour import CyclicBehaviour
 from spade.message import Message
 from spade.template import Template
 
+from dataclasses import dataclass
+import calendar
+import time
+
+from constants import DEFAULT_HOST
+from messages.check_offers import CheckOffers
+from messages.consolidated_offers import ConsolidatedOffers
+from messages.reservation_request import RequestReservation
+from messages.response import ReservationResponse
+
+RECEIVE_TIMEOUT=1
 
 class User(Agent):
-    class CheckUserBehaviour(CyclicBehaviour):
+    def __init__(self, jid, password):
+        super().__init__(jid, password)
+        self.coordinators_jids = None
+        self.pending_reservation = None     
+        self.active_reservations = []     
+
+    def _validate_timeslot(self, time_start, time_stop):
+        current_time = calendar.timegm(time.gmtime()) # Get current timestamp in UTC timezone
+        if not (time_start < time_stop and time_start > current_time):
+            raise ValueError("Invalid timeslot")
+    
+    def _validate_location(self, x, y):
+        if not (x >= 0 and y >= 0):
+            raise ValueError("Invalid location")
+        
+    def _prepare_consolidated_offers_template(self):
+        template = Template()
+        template.to = f"{self.jid}@{DEFAULT_HOST}"
+        template.set_metadata("performative", "inform")
+        template.set_metadata("action", "consolidated-offers")
+        return template
+    
+    def _prepare_reservation_response_template(self):
+        template = Template()
+        template.to = f"{self.jid}@{DEFAULT_HOST}"
+        template.set_metadata("performative", "inform")
+        template.set_metadata("action", "reservation-response")
+        return template
+
+    def add_coordinator(self, coordinator_jid):
+        self.coordinators_jids.append(coordinator_jid)
+
+    async def request_parking_offers(self, x, y, time_start, time_stop):
+
+        self._validate_location(x, y)
+        self._validate_timeslot(time_start, time_stop)
+
+        self.pending_reservation = Reservation(x, y, time_start, time_stop)
+
+        check_offers_msg = CheckOffers(
+            x=x,
+            y=y,
+            time_start=time_start,
+            time_stop=time_stop
+        )
+        for coordinator_jid in self.coordinators_jids:
+            to_send = Message(
+                to=coordinator_jid,
+                body=json.dumps(check_offers_msg.dict()),
+                metadata={"performative": "query-ref", "action": "check-offers"}
+            )
+            await self.send(to_send)
+
+    def choose_parking_offer(self, offers: List[dict]):
+        #TODO implement logic of choosing an offer
+        return offers[0]["parking_id"] if offers else None
+
+    class MakeReservation(CyclicBehaviour):
```

Comment:
Is this a `CyclicBehaviour` or an `OneShotBehaviour`?

Reply:
As I understand from documentation, OneShotBehaviour could be realized only one time in agent life, so if we would like to make multiple reservations, we should have this behaviour as CyclicBehaviour